File: backend/src/main/java/br/ufpr/sept/so2/modules/bff/infrastructure/AlunoIdentidadeQueryAdapter.java
package br.ufpr.sept.so2.modules.bff.infrastructure;

import br.ufpr.sept.so2.modules.academico.application.ports.AlunoRepository;
import br.ufpr.sept.so2.modules.academico.application.ports.CursoRepository;
import br.ufpr.sept.so2.modules.academico.domain.Aluno;
import br.ufpr.sept.so2.modules.bff.application.ports.AlunoIdentidadeQueryPort;
import br.ufpr.sept.so2.modules.iam.application.ports.UsuarioRepository;
import br.ufpr.sept.so2.modules.iam.domain.Usuario;
import br.ufpr.sept.so2.shared.domain.exception.RecursoNaoEncontradoException;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;
import java.util.UUID;

@Component
public class AlunoIdentidadeQueryAdapter implements AlunoIdentidadeQueryPort {

    private final UsuarioRepository usuarioRepository;
    private final AlunoRepository alunoRepository;
    private final CursoRepository cursoRepository;

    public AlunoIdentidadeQueryAdapter(
            UsuarioRepository usuarioRepository,
            AlunoRepository alunoRepository,
            CursoRepository cursoRepository
    ) {
        this.usuarioRepository = usuarioRepository;
        this.alunoRepository = alunoRepository;
        this.cursoRepository = cursoRepository;
    }

    @Override
    @Transactional(readOnly = true)
    public AlunoIdentidade consultar(UUID usuarioId) {
        Usuario usuario = usuarioRepository.findById(usuarioId)
                .orElseThrow(() -> new RecursoNaoEncontradoException("Usuário autenticado não encontrado."));
        Optional<Aluno> aluno = Optional.empty();
        if (usuario.getGrr() != null) {
            aluno = alunoRepository.findByGrr(usuario.getGrr().getValue());
        }
        if (aluno.isEmpty() && usuario.getEmailInstitucional() != null) {
            aluno = alunoRepository.findByEmailInstitucional(usuario.getEmailInstitucional().getValue());
        }
        String nome = aluno.map(AlunoIdentidadeQueryAdapter::nomeExibicao)
                .orElseGet(() -> nomeDoEmail(usuario.getEmailInstitucional().getValue()));
        String curso = aluno
                .flatMap(encontrado -> cursoRepository.findById(encontrado.getIdCurso()))
                .map(item -> item.getSigla() != null && !item.getSigla().isBlank() ? item.getSigla() : item.getNome())
                .orElse(null);
        return new AlunoIdentidade(nome, curso);
    }

    private static String nomeExibicao(Aluno aluno) {
        if (aluno.getNomeSocial() != null && !aluno.getNomeSocial().isBlank()) {
            return aluno.getNomeSocial().trim();
        }
        return aluno.getNome();
    }

    private static String nomeDoEmail(String email) {
        String local = email.substring(0, email.indexOf('@')).replace('.', ' ').trim();
        if (local.isEmpty()) {
            return "Aluno";
        }
        String[] partes = local.split("\\s+");
        StringBuilder builder = new StringBuilder();
        for (String parte : partes) {
            if (parte.isEmpty()) {
                continue;
            }
            if (!builder.isEmpty()) {
                builder.append(' ');
            }
            builder.append(Character.toUpperCase(parte.charAt(0)));
            if (parte.length() > 1) {
                builder.append(parte.substring(1));
            }
        }
        return builder.toString();
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/modules/bff/infrastructure/AlunoIdentidadeQueryAdapter.kt
package br.ufpr.sept.so2.modules.bff.infrastructure

import br.ufpr.sept.so2.modules.academico.application.ports.AlunoRepository
import br.ufpr.sept.so2.modules.academico.application.ports.CursoRepository
import br.ufpr.sept.so2.modules.academico.domain.Aluno
import br.ufpr.sept.so2.modules.bff.application.ports.AlunoIdentidadeQueryPort
import br.ufpr.sept.so2.modules.bff.application.ports.AlunoIdentidadeQueryPort.AlunoIdentidade
import br.ufpr.sept.so2.modules.iam.application.ports.UsuarioRepository
import br.ufpr.sept.so2.shared.domain.exception.RecursoNaoEncontradoException
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.util.Optional
import java.util.UUID

@Component
class AlunoIdentidadeQueryAdapter(
    private val usuarioRepository: UsuarioRepository,
    private val alunoRepository: AlunoRepository,
    private val cursoRepository: CursoRepository,
) : AlunoIdentidadeQueryPort {

    @Transactional(readOnly = true)
    override fun consultar(usuarioId: UUID): AlunoIdentidade {
        val usuario = usuarioRepository.findById(usuarioId)
            .orElseThrow { RecursoNaoEncontradoException("Usuário autenticado não encontrado.") }
        var aluno: Optional<Aluno> = Optional.empty()
        val grr = usuario.grr
        if (grr != null) {
            aluno = alunoRepository.findByGrr(grr.value)
        }
        val emailInstitucional = usuario.emailInstitucional
        if (aluno.isEmpty && emailInstitucional != null) {
            aluno = alunoRepository.findByEmailInstitucional(emailInstitucional.value)
        }
        val nome = aluno.map(::nomeExibicao).orElseGet { nomeDoEmail(emailInstitucional.value) }
        val curso = aluno
            .flatMap { encontrado -> cursoRepository.findById(encontrado.idCurso) }
            .map { item -> if (item.sigla.isNotBlank()) item.sigla else item.nome }
            .orElse(null)
        return AlunoIdentidade(nome, curso)
    }

    companion object {
        private fun nomeExibicao(aluno: Aluno): String {
            val social = aluno.nomeSocial
            if (!social.isNullOrBlank()) {
                return social.trim()
            }
            return aluno.nome
        }

        private fun nomeDoEmail(email: String): String {
            val local = email.substring(0, email.indexOf('@')).replace('.', ' ').trim()
            if (local.isEmpty()) {
                return "Aluno"
            }
            return local.split(Regex("\\s+"))
                .filter { it.isNotEmpty() }
                .joinToString(" ") { parte ->
                    parte.replaceFirstChar { it.uppercaseChar() }
                }
        }
    }
}
